Initial pull request to get the project started.

chbrown13 opened the pull request:
Added an initial setup.py and partition_disk job as an example.

However, much of this could be up for change depending on the discussion had by the entire team.
One of the primary discussions being how we want to run these tasks:
- ansible-runner --hosts test.redhat.com test2.redhat.com <task>
- ansible-satellite6 <task>
- ansible-satellite6 <task> --hosts test.redhat.com test2.redhat.com

Commit:
Initial pull request to get the project started.

Added an initial setup.py and partition_disk job as an example.

However, much of this could be up for change depending on the discussion had by the entire team.
One of the primary discussions being how we want to run these tasks:
- ansible-runner --hosts test.redhat.com test2.redhat.com <task>
- ansible-satellite6 <task>
- ansible-satellite6 <task> --hosts test.redhat.com test2.redhat.com

bkearney commented on ansible-satellite6/ansible_runner/project/partition_disk.yml:
```diff
@@ -0,0 +1,19 @@
+    - debug: msg="{{ disk_free.stdout }}"
+    
+    - shell: |
+        umount /home
```
Why not use https://docs.ansible.com/ansible/latest/modules/parted_module.html

JacobCallahan replied:
Great feedback. Chris is fairly new to ansible. So, like myself, he probably isn't aware of all the modules available to make some of these things easier.
@chbrown13 This would be a great module to look into!

chbrown13 replied:
Ok cool, will take a look. Thanks @bkearney

pgagne replied:
Might want to look at these:

https://docs.ansible.com/ansible/latest/modules/lvg_module.html?highlight=logical%20volumes
https://docs.ansible.com/ansible/latest/modules/lvol_module.html?highlight=logical%20volumes

Since we actually aren't partitioning the disk, but removing and resizing logical volumes. We might want to come up with a better term for this too.

flatfender replied:
Also I think we want to make sure we have a directory structure that mimics what galaxy would use in most cases.  Because that will make it easier is we want to contribute upstream to galaxy as well.  "ansible-galaxy init --offline -p roles rolename" is very useful as it will create a empty galaxy directory structure.

JacobCallahan replied:
@flatfender that change will be made, if needed, when we discuss our plans for this repo on Friday. The current structure is to be compatible with ansible-runner.

chris1984 commented on ansible-satellite6/ansible_runner/project/partition_disk.yml:
```diff
@@ -0,0 +1,19 @@
+        umount /home
+        lvremove -f /dev/mapper/*home
+        sed -i '/\/home/d' /etc/fstab
+        lvresize -f -l +100%FREE /dev/mapper/*root
```
we can `resize2fs` at the same time of resizing the volume by doing `-r`. I am sure the module the others are talking about have an option to resize during the operation but for shell in general you can do `-r` at time of resize.

chris1984 commented on ansible-satellite6/ansible_runner/project/partition_disk.yml:
```diff
@@ -0,0 +1,19 @@
+    - shell: |
+        umount /home
+        lvremove -f /dev/mapper/*home
+        sed -i '/\/home/d' /etc/fstab
```
A good check after editing the fstab is to run the command `mount all` to see if there is any errors, the modules the others are talking about prob have ways to do that already but in general always good to check before a reboot since the machine will come up in emergency mode if something in fstab is missing/incorrect.

chris1984 reviewed (commented):
Left a few comments.

jhutar commented:
Maybe could all this be a **role**? We could submit it to Galaxy so Satellite customers can easily benefit from it as well (in Sat you can easily assign roles, not plays/playbooks).

If we make all functions a nicely defined role, it would be pluggable into many projects most of us have (you can do `ansible-galaxy install -r ansible-requirements.txt`, similar to pip).

ehelms commented:
I agree roles are good, I wouldn't be thinking galaxy from the start as that can complicate and slow the pace as users can just as easily git clone as galaxy install IMO.

I would also point you at considering linting out of the gates -- https://github.com/willthames/ansible-lint

jhutar commented:
Why I mentioned Galaxy was that there are other projects which could benefit from the roles. But I have no idea how hard is to get role to Galaxy: if it is as simple as for Pip, I would be all for it. If this happens, we easilly drop same role satperf have:

https://github.com/redhat-performance/satperf/tree/master/playbooks/common/roles/remove-home-extend-root/tasks

And yes, I'm all for linting (from beginning).

jhutar commented:
And, once this is a role, do we need the python script at all? Why it is better than just using Ansible? I think it could lead to hiding some logic/functionality in Python and not in the roles, which would make reusing roles harder.

Commit:
Updates based on feedback

chbrown13 commented:
**Before:**

[root@ibm-x3550m3-10 ~]# cat /etc/fstab 

#
# /etc/fstab
# Created by anaconda on Wed Jul 11 16:16:57 2018
#
# Accessible filesystems, by reference, are maintained under '/dev/disk'
# See man pages fstab(5), findfs(8), mount(8) and/or blkid(8) for more info
#
/dev/mapper/rhel_ibm--x3550m3--10-root /                       xfs     defaults        0 0
UUID=6fe454fb-d610-46a1-a8e2-1ddf40920063 /boot                   xfs     defaults        0 0
/dev/mapper/rhel_ibm--x3550m3--10-home /home                   xfs     defaults        0 0
/dev/mapper/rhel_ibm--x3550m3--10-swap swap                    swap    defaults        0 0
[root@ibm-x3550m3-10 ~]# lvdisplay
  --- Logical volume ---
  LV Path                /dev/rhel_ibm-x3550m3-10/swap
  LV Name                swap
  VG Name                rhel_ibm-x3550m3-10
  LV UUID                B2LHr8-Vt7f-a6iV-tBFh-vpdX-DbBx-8Vv9Nn
  LV Write Access        read/write
  LV Creation host, time ibm-x3550m3-10.lab.eng.brq.redhat.com, 2018-07-11 18:16:54 +0200
  LV Status              available
  # open                 2
  LV Size                11.81 GiB
  Current LE             3024
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           253:1
   
  --- Logical volume ---
  LV Path                /dev/rhel_ibm-x3550m3-10/home
  LV Name                home
  VG Name                rhel_ibm-x3550m3-10
  LV UUID                cKjFcp-TZID-AuET-4nuW-9S7b-KQcC-YhmeVc
  LV Write Access        read/write
  LV Creation host, time ibm-x3550m3-10.lab.eng.brq.redhat.com, 2018-07-11 18:16:55 +0200
  LV Status              available
  # open                 1
  LV Size                401.91 GiB
  Current LE             102889
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           253:2
   
  --- Logical volume ---
  LV Path                /dev/rhel_ibm-x3550m3-10/root
  LV Name                root
  VG Name                rhel_ibm-x3550m3-10
  LV UUID                3KfkZr-8oI2-if0y-LvOl-nsWD-iTVZ-lkY0B9
  LV Write Access        read/write
  LV Creation host, time ibm-x3550m3-10.lab.eng.brq.redhat.com, 2018-07-11 18:16:55 +0200
  LV Status              available
  # open                 1
  LV Size                50.00 GiB
  Current LE             12800
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           253:0




$ ansible-playbook satellite.yml --inventory /etc/ansible/hosts --ask-pass
SSH password: 

PLAY [all] ********************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : disk-free] *********************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : debug] *************************************************************************************************************************************************************************************
ok: [ibm-x3550m3-10.lab.eng.brq.redhat.com] => {
    "msg": "/home"
}

TASK [partition-disk : group-name] ********************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : debug-vg] **********************************************************************************************************************************************************************************
ok: [ibm-x3550m3-10.lab.eng.brq.redhat.com] => {
    "msg": {
        "changed": true, 
        "cmd": "vgs --noheadings -o vg_name | tr -d '  '", 
        "delta": "0:00:00.018948", 
        "end": "2018-07-11 18:34:47.915561", 
        "failed": false, 
        "rc": 0, 
        "start": "2018-07-11 18:34:47.896613", 
        "stderr": "", 
        "stderr_lines": [], 
        "stdout": "rhel_ibm-x3550m3-10", 
        "stdout_lines": [
            "rhel_ibm-x3550m3-10"
        ]
    }
}

TASK [partition-disk : unmount-home] ******************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : remove-logical-volume] *********************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : stream-editor3] ****************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : mount-all] *********************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : resize-logical-volume] *********************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : kernel-release] ****************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : increase-size] *****************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

PLAY RECAP ********************************************************************************************************************************************************************************************************
ibm-x3550m3-10.lab.eng.brq.redhat.com : ok=12   changed=9    unreachable=0    failed=0

chbrown13 commented:
**After:**

[root@ibm-x3550m3-10 ~]# cat /etc/fstab 

#
# /etc/fstab
# Created by anaconda on Wed Jul 11 16:16:57 2018
#
# Accessible filesystems, by reference, are maintained under '/dev/disk'
# See man pages fstab(5), findfs(8), mount(8) and/or blkid(8) for more info
#
/dev/mapper/rhel_ibm--x3550m3--10-root /                       xfs     defaults        0 0
UUID=6fe454fb-d610-46a1-a8e2-1ddf40920063 /boot                   xfs     defaults        0 0
/dev/mapper/rhel_ibm--x3550m3--10-swap swap                    swap    defaults        0 0
[root@ibm-x3550m3-10 ~]# lvdisplay
  --- Logical volume ---
  LV Path                /dev/rhel_ibm-x3550m3-10/swap
  LV Name                swap
  VG Name                rhel_ibm-x3550m3-10
  LV UUID                B2LHr8-Vt7f-a6iV-tBFh-vpdX-DbBx-8Vv9Nn
  LV Write Access        read/write
  LV Creation host, time ibm-x3550m3-10.lab.eng.brq.redhat.com, 2018-07-11 18:16:54 +0200
  LV Status              available
  # open                 2
  LV Size                11.81 GiB
  Current LE             3024
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           253:1
   
  --- Logical volume ---
  LV Path                /dev/rhel_ibm-x3550m3-10/root
  LV Name                root
  VG Name                rhel_ibm-x3550m3-10
  LV UUID                3KfkZr-8oI2-if0y-LvOl-nsWD-iTVZ-lkY0B9
  LV Write Access        read/write
  LV Creation host, time ibm-x3550m3-10.lab.eng.brq.redhat.com, 2018-07-11 18:16:55 +0200
  LV Status              available
  # open                 1
  LV Size                451.91 GiB
  Current LE             115690
  Segments               2
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           253:0




$ ansible-playbook satellite.yml --inventory /etc/ansible/hosts
SSH password: 

PLAY [all] ********************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : disk-free] *********************************************************************************************************************************************************************************
changed: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : debug] *************************************************************************************************************************************************************************************
ok: [ibm-x3550m3-10.lab.eng.brq.redhat.com] => {
    "msg": "/"
}

TASK [partition-disk : group-name] ********************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : debug-vg] **********************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : unmount-home] ******************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : remove-logical-volume] *********************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : stream-editor3] ****************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : mount-all] *********************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : resize-logical-volume] *********************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : kernel-release] ****************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

TASK [partition-disk : increase-size] *****************************************************************************************************************************************************************************
skipping: [ibm-x3550m3-10.lab.eng.brq.redhat.com]

PLAY RECAP ********************************************************************************************************************************************************************************************************
ibm-x3550m3-10.lab.eng.brq.redhat.com : ok=3    changed=1    unreachable=0    failed=0

JacobCallahan commented:
@chbrown13 with the switch to pure ansible, we can remove the python parts of this. That includes the __init__.py and setup.py

Commit:
Removing python files

kbidarkar commented on ansible-satellite6/roles/partition-disk/tasks/main.yml:
```diff
@@ -0,0 +1,61 @@
+    register: kernel
+
+  - name: increase-size
+    shell: xfs_growfs / && mount / -o inode64,remount
```
Instead of shell command, we need to make use of `filesystem` module.

https://docs.ansible.com/ansible/latest/modules/filesystem_module.html

kbidarkar commented:
@chbrown13, thanks for getting this started, but for the moment would just mark this as `DO NOT MERGE`.

ehelms commented on .gitignore:
```diff
@@ -53,5 +53,5 @@ coverage.xml
-# ansible-runner
-ansible_runner/artifacts/*
+# Ansible
+*.retry
```
I find these can clutter the workspace, you can turn them off or set a global configuration spot for them e.g. ~/.cache/retries

ehelms commented on ansible-satellite6/roles/partition-disk/tasks/main.yml:
```diff
@@ -0,0 +1,61 @@
+---
```
Is it intentional to have roles nested inside a nested ansible-satellite6 directory?

chbrown13 replied:
Yes, it was intentional.

ehelms commented on ansible-satellite6/satellite.yml:
```diff
@@ -0,0 +1,5 @@
+---
+- hosts: all
+  remote_user: root
```
I think it can be better to avoid setting this in the playbook and rather set that at the inventory level.

kbidarkar commented:
@chbrown13, Also, just realized, please delete the test output it has a lot of sensitive stuff on GitHub.

kbidarkar commented on ansible-satellite6/roles/partition-disk/tasks/main.yml:
```diff
@@ -0,0 +1,61 @@
+
+  - name: remove-logical-volume
+    lvol:
+      lv: /dev/"{{vg.stdout}}"/home
```
Also, it would be ideal if we go with using ansible variables along with `with_dict` option to populate the values of `lv` and `vg`.

```
qe-sat6-feature-rhel7 | SUCCESS => {
    "ansible_facts": {
        "ansible_lvm": {
            "lvs": {
                "lv_root": {
                    "size_g": "193.66", 
                    "vg": "vg_dhcp201"
                }, 
                "lv_swap": {
                    "size_g": "5.84", 
                    "vg": "vg_dhcp201"
                }
            }, 
            "pvs": {
                "/dev/vda2": {
                    "free_g": "0", 
                    "size_g": "199.51", 
                    "vg": "vg_dhcp201"
                }
            }, 
            "vgs": {
                "vg_dhcp201": {
                    "free_g": "0", 
                    "num_lvs": "2", 
                    "num_pvs": "1", 
                    "size_g": "199.51"
                }
            }
        }
    }, 
    "changed": false
```

Similar to below stuff,

```
- hosts: test
  vars:
    lv:
      vol1:
        vg: cl
        size: 1g
      vol2:
        vg: cl
        size: 5g
      vol3:
        vg: cl
        size: 1g
  tasks:
  - name: Debug mode to see if ansible itself passes variables
    debug: msg="{{ item.key }} --> {{ item.value.vg }} --> {{ item.value.size }}"
    with_dict: "{{ lv }}"

  - name: Test lvol module to build an example lv layout
    lvol:
      vg: "{{ item.value.vg }}"
      lv: "{{ item.key }}"
      size: "{{ item.value.size }}"
      active: yes
      state: present
    with_dict: "{{ lv }}"
```

kbidarkar commented:
Will close this PR and as mentioned in the mail, will work towards the directory structure, ansible rules and standard and raise PR's for it. Also will be sharing a mojo doc with the entire plan.
